libraries/AP_WindVane/AP_WindVane.h

```diff
@@ -0,0 +1,109 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Param/AP_Param.h>
+#include <AP_Math/AP_Math.h>
+
+class AP_WindVane
+{
+
+public:
+
+    enum WindVaneType {
+        WINDVANE_NONE       = 0,
+        WINDVANE_HOME_HEADING = 1,
+        WINDVANE_PWM_PIN    = 2,
+        WINDVANE_ANALOG_PIN = 3,
+        WINDVANE_SITL       = 10
+    };
+
+    AP_WindVane();
+
+    /* Do not allow copies */
+    AP_WindVane(const AP_WindVane &other) = delete;
+    AP_WindVane &operator=(const AP_WindVane&) = delete;
+
+    // destructor
+    ~AP_WindVane(void);
+
+    static AP_WindVane *get_instance();
+
+    // return true if wind vane is enabled
+    bool enabled() const;
+
+    // Initialize the Wind Vane object and prepare it for use
+    void init();
+
+    // update wind vane
+    void update();
+
+    // get the apparent wind direction in radians, 0 = head to wind
+    float get_apparent_wind_direction_rad() const { return _direction_apparent; }
```

Comment:
Why is this not expressed to true north? Mostly just curious because from a data staleness perspective you might only call update on this at a lower rate, but using a stale wind direction is probably better then seeing a reported sawtooth on wind direction. At low speeds/turns I admit this probably wont ever be a problem.

Reply:
the wind direction is relative to the vehicle's heading because this is what is used to set the mainsail (i.e. we pull in the sail when the apparent wind is forward, let it out when it is coming from the side) and also because this is the frame that the sensor reports in (it is normally a windvane attached to the top of the mast).  In a future PR the AP_WindVane will estimate the absolute wind direction which is important for knowing when to tack.

Reply:
.. but I see what you are saying and it has crossed my mind as well.  To make sure we are on the same page, I guess what you're suggesting is that we should record the apparent wind direction in the earth frame (i.e. relative to North) and then when the caller requests the apparent wind direction (and they need it in body frame) we convert it back.  This would remove the saw-tooth caused by the vehicle rotating between updates.  I think this is a great idea.. we will add this.

I should add that it crossed my mind because of the filtering which adds a delay of course and makes the problem even worse.  i.e. we really want to apply the filter to a value that is not affected by the rotation of the vehicle.